src/gui/tagfiltermodel.h

```diff
@@ -0,0 +1,85 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#ifndef TAGFILTERMODEL_H
+#define TAGFILTERMODEL_H
+
+#include <QAbstractListModel>
+#include <QHash>
+#include <QModelIndex>
+#include <QSet>
+#include <QVector>
+
+namespace BitTorrent
+{
+    class TorrentHandle;
+}
+
+class TagModelItem;
+
+class TagFilterModel: public QAbstractListModel
+{
+    Q_OBJECT
+
+public:
+    explicit TagFilterModel(QObject *parent = nullptr);
+    ~TagFilterModel();
+
+    static bool isSpecialItem(const QModelIndex &index);
+
+    QVariant data(const QModelIndex &index, int role = Qt::DisplayRole) const override;
+    Qt::ItemFlags flags(const QModelIndex &index) const override;
+    QVariant headerData(int section, Qt::Orientation orientation, int role) const override;
+    QModelIndex index(int row, int column, const QModelIndex &parent = QModelIndex()) const override;
+    int rowCount(const QModelIndex &parent = QModelIndex()) const override;
+
+    QModelIndex index(const QString &tag) const;
+    QString tag(const QModelIndex &index) const;
+
+private slots:
+    void tagAdded(const QString &tag);
+    void tagRemoved(const QString &tag);
+    void torrentTagAdded(BitTorrent::TorrentHandle *const torrent, const QString &tag);
+    void torrentTagRemoved(BitTorrent::TorrentHandle *const, const QString &tag);
+    void torrentAdded(BitTorrent::TorrentHandle *const torrent);
+    void torrentAboutToBeRemoved(BitTorrent::TorrentHandle *const torrent);
+
+private:
+    static QString tagDisplayName(const QString &tag);
+
+    void populate();
+    void addItemToModel(TagModelItem *item);
+    void removeItemFromModel(TagModelItem *item);
+    QModelIndex index(const TagModelItem *item) const;
+    TagModelItem *findItem(const QString &tag) const;
+    QVector<TagModelItem *> findItems(const QSet<QString> &tags) const;
+
+    QList<TagModelItem *> m_tagItems;  // Index corresponds to its row
```

Comment:
@glassez, we got less complex code already. And yes, I avoid pointers when possible not only because C++ provides better ways to manage objects (but that's the main reason), but also because of data locality, which is important for qBt too, not because it needs all the performance of the CPU cache, but because programs like this one are active in background and provokes cache misses for other processes.